In [75]:
import numpy as np
import pandas as pd

from configparser import ConfigParser
config = ConfigParser()
# create your own config.ini in root of project folder to store project configurations
config.read('config.ini')


pathfile = config.get('main', 'dirty_csv')       

df = pd.read_csv(pathfile, nrows=10**6)
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 30)

df.head(20)



ActualElapsedTime  AirTime  ArrDelay  ArrTime  CRSArrTime  CRSDepTime  \
0                53.0     32.0      -8.0   1642.0        1650        1545   
1               164.0    155.0     -11.0   1754.0        1805        1610   
2                60.0      NaN      15.0   2005.0        1950        1850   
3                51.0      NaN      -5.0   1818.0        1823        1728   
4                45.0     29.0       2.0   1120.0        1118        1030   
5                49.0     37.0       2.0   1137.0        1135        1048   
6                61.0     40.0      -3.0   1537.0        1540        1440   
7               150.0    126.0     -19.0   2015.0        2034        1745   
8               115.0    103.0      -5.0    735.0         740         640   
9                 NaN      NaN       NaN      NaN         730         605   
10              102.0      NaN      -4.0   1603.0        1607        1320   
11               87.0      NaN      -8.0   1547.0        1555        1322   
12              170.0    135.0      10.0   1051.0        1041         800   
13              100.0     81.0      -2.0   1220.0        1222        1045   
14               60.0      NaN      -5.0   2040.0        2045        1940   
15               79.0     70.0     -16.0   1444.0        1500        1220   
16              174.0      NaN      43.0   1213.0        1130        1031   
17              133.0    117.0      40.0   1429.0        1349        1044   
18               33.0     22.0      -3.0   1916.0        1919        1845   
19               57.0     31.0      -9.0   1659.0        1708        1612   

    CRSElapsedTime CancellationCode  Cancelled  CarrierDelay  DayOfWeek  \
0             65.0              NaN          0           NaN          4   
1            175.0              NaN          0           NaN          4   
2             60.0              NaN          0           NaN          5   
3             55.0              NaN          0           NaN          4   
4             48.0              NaN          0           0.0          1   
5             47.0              NaN          0           NaN          4   
6             60.0              NaN          0           NaN          7   
7            169.0              NaN          0           NaN          4   
8            120.0              NaN          0           NaN          2   
9             85.0                B          1           0.0          7   
10           107.0              NaN          0           NaN          2   
11            93.0              NaN          0           NaN          3   
12           161.0              NaN          0           NaN          4   
13            97.0              NaN          0           0.0          1   
14            65.0              NaN          0           NaN          6   
15           100.0              NaN          0           NaN          7   
16           179.0              NaN          0           NaN          5   
17           125.0              NaN          0           NaN          4   
18            34.0              NaN          0           0.0          2   
19            56.0              NaN          0           0.0          5   

    DayofMonth  DepDelay  DepTime Dest  Distance  Diverted  FlightNum  \
0           10       4.0   1549.0  PIT     205.0         0        209   
1            2       0.0   1610.0  MCI    1072.0         0        109   
2           10      15.0   1905.0  CLT     227.0         0       1276   
3           28      -1.0   1727.0  BNA     200.0         0        961   
4           19       5.0   1035.0  CMH     116.0         0       5873   
5            2       0.0   1048.0  CLT     156.0         0        353   
6           20      -4.0   1436.0  LAW     140.0         0       3281   
7           15       0.0   1745.0  ATL     903.0         0       1521   
8           16       0.0    640.0  SEA     689.0         0       1678   
9           10       NaN      NaN  SLC     391.0         0       4117   
10     

# ----------------------cleaning comments----------------------
- Total number of records > 120 million
- remove all records with more than certain number of NaN values?
- At least remove all records with NaNs for assumed FDS except for the ones where NaN has to be transformed to something (will see this for CarrierDelay below)?

- CarrierDelay has 0.0 or NaN for when there is no delay? Turn NaN into 0?
    - Same goes for every other Delay column. Which delay columns do we want to keep?
- CancellationCode has letter for when Cancelled = 1 and NaN for when not cancelled
    - Change NaNs in CancellationCode to unique not-cancelled code?

- Remove records where TailNum = 0? Is normally several numbers and a letter. Or keep for soft functional dependency?
 

In [10]:
df.describe()

ActualElapsedTime        AirTime       ArrDelay        ArrTime  \
count      978757.000000  681360.000000  978757.000000  978783.000000   
mean          120.218606     102.854060       7.051868    1492.796174   
std            68.704049      71.022941      30.809895     498.580434   
min          -525.000000   -1419.000000   -1410.000000       1.000000   
25%            70.000000      54.000000      -7.000000    1115.000000   
50%           102.000000      84.000000       0.000000    1521.000000   
75%           151.000000     133.000000      11.000000    1916.000000   
max          1812.000000    1570.000000    1435.000000    2717.000000   

           CRSArrTime      CRSDepTime  CRSElapsedTime       Cancelled  \
count  1000000.000000  1000000.000000   999791.000000  1000000.000000   
mean      1490.388786     1333.926632      121.124092        0.018890   
std        493.228099      476.174529       68.105616        0.136137   
min          0.000000        0.000000      -30.000000        0.000000   
25%       1115.000000      928.000000       71.000000        0.000000   
50%       1520.000000     1326.000000      102.000000        0.000000   
75%       1912.000000     1727.000000      152.000000        0.000000   
max       2400.000000     2359.000000      660.000000        1.000000   

        CarrierDelay       DayOfWeek      DayofMonth       DepDelay  \
count  276290.000000  1000000.000000  1000000.000000  981117.000000   
mean        3.708991        3.941351       15.735314       8.158946   
std        19.515270        1.990578        8.793015      28.673504   
min         0.000000        1.000000        1.000000   -1410.000000   
25%         0.000000        2.000000        8.000000      -2.000000   
50%         0.000000        4.000000       16.000000       0.000000   
75%         0.000000        6.000000       23.000000       6.000000   
max      1126.000000        7.000000       31.000000    1439.000000   

             DepTime       Distance        Diverted       FlightNum  \
count  981117.000000  998310.000000  1000000.000000  1000000.000000   
mean     1347.962749     702.446892        0.002353     1364.394253   
std       476.983626     551.751271        0.048451     1403.895626   
min         1.000000      11.000000        0.000000        1.000000   
25%       933.000000     307.000000        0.000000      450.000000   
50%      1333.000000     545.000000        0.000000      945.000000   
75%      1737.000000     937.000000        0.000000     1711.000000   
max      2635.000000    4983.000000        1.000000     9619.000000   

       LateAircraftDelay           Month       NASDelay  SecurityDelay  \
count      276290.000000  1000000.000000  276290.000000  276290.000000   
mean            4.715715        6.556942       4.099834       0.024033   
std            20.071980        3.447362      16.662616       0.954795   
min             0.000000        1.000000      -1.000000       0.000000   
25%             0.000000        4.000000       0.000000       0.000000   
50%             0.000000        7.000000       0.000000       0.000000   
75%             0.000000       10.000000       0.000000       0.000000   
max           613.000000       12.000000     875.000000     128.000000   

              TaxiIn        TaxiOut   WeatherDelay            Year  
count  696728.000000  696823.000000  276290.000000  1000000.000000  
mean        6.476692      15.218328       0.760534     1998.614886  
std        22.342821      11.073621       8.904026        6.224338  
min         0.000000       0.000000       0.000000     1987.000000  
25%         4.000000      10.000000       0.000000     1993.000000  
50%         5.000000      13.000000       0.000000     1999.000000  
75%         7.000000      18.000000       0.000000     2004.000000  
max      1460.000000    1435.000000     958.000000     2008.000000

# check strong FD (TailNum) -> (UniqueCarrier)

In [76]:
# ---------------- check strong FD (TailNum) -> (UniqueCarrier)---------------
# dependency is not a bi-implication
df_uniq_carr = df[['TailNum','UniqueCarrier']]
df_uniq_carr = df_uniq_carr.sort_values(by=['TailNum'])
df_uniq_carr.head(20)
# describe for for example unique value counts
# df_uniq_carr.describe()

TailNum UniqueCarrier
864543  -N037M            MQ
425218  -N037M            MQ
860027  -N037M            MQ
844197  -N047M            MQ
994545  -N047M            MQ
454325  -N047M            MQ
119211  -N047M            MQ
798038  -N047M            MQ
162424  -N047M            MQ
690297  -N107D            DL
133494  -N107D            DL
235247  -N107D            DL
589633  -N107D            DL
576788  -N107D            DL
554885  -N107D            DL
877855  -N107D            DL
423141  -N107D            DL
788988  -N107D            DL
363479  -N107D            DL
154337  -N107D            DL

In [88]:
# check if dependency also works other way around. Seems not.
# Just pick one uniquecarrier value and filter NaNs on TailNum side
df_uniq_carr_filt = df_uniq_carr[(df_uniq_carr['UniqueCarrier'] == '9E') & (df_uniq_carr['TailNum'].notnull())]
# df_uniq_carr_filt.describe()
df_uniq_carr_filt.tail(50)

TailNum UniqueCarrier
2146    91709E            9E
254277  91709E            9E
403016  91709E            9E
910415  91709E            9E
138319  91709E            9E
70986   91709E            9E
42137   91709E            9E
222172  91709E            9E
804150  91709E            9E
160478  91709E            9E
266069  91709E            9E
823969  91769E            9E
614053  91769E            9E
209094  91769E            9E
767254  91769E            9E
508541  91769E            9E
558020  91769E            9E
396018  91819E            9E
151978  91819E            9E
518085  91819E            9E
275046  91819E            9E
826382  91819E            9E
405782  91819E            9E
5434    91819E            9E
713639  91819E            9E
148730  91819E            9E
864068  91819E            9E
861238  91869E            9E
756367  91869E            9E
762445  91869E            9E
652971  91879E            9E
286468  91879E            9E
829881  91879E            9E
514062  91879E            9E
26143   91879E            9E
791961  91879E            9E
812320  96009E            9E
514249  96029E            9E
671802  96049E            9E
109619  96049E            9E
520839  96059E            9E
926572  96059E            9E
317346  96069E            9E
744354  96069E            9E
920265  96079E            9E
772796  96079E            9E
933423  96079E            9E
75648   96079E            9E
23411   96079E            9E
869768  96079E            9E

# check strong FD (Origin, Dest) -> (Distance)

In [112]:
df_origin_dest = df[['Origin', 'Dest', 'Distance']]
# df_origin_dest = df_origin_dest.sort_values(by=['Origin', 'Dest'])
df_origin_dest = df_origin_dest.sort_values(by=['Distance','Origin', 'Dest'])
# filter NaN
df_origin_dest[df_origin_dest['Distance'].notnull()].tail(500)
# compare amount of different groups of [origin, dest, distance] (full FD) with groups of [origin, dest] (left side FD)
# should be same number of groups for strong dependency
df_origin_dest.groupby(['Origin', 'Dest', 'Distance']).size().reset_index().rename(columns={0:'count'}).count()

Origin      6775
Dest        6775
Distance    6775
count       6775
dtype: int64

In [113]:
# counts for only left hand side FD
df_origin_dest.groupby(['Origin', 'Dest']).size().reset_index().rename(columns={0:'count'}).count()

Origin    6775
Dest      6775
count     6775
dtype: int64

# check Soft FD (CRSDepTime, DepTime) -> (DepDelay)

In [71]:
df_crsdep_dep = df[['CRSDepTime', 'DepTime', 'DepDelay']]
df_crsdep_dep = df_crsdep_dep.sort_values(by=['CRSDepTime', 'DepTime'])
# df_crsdep_dep.head(500)
df_crsdep_dep.groupby(['CRSDepTime', 'DepTime', 'DepDelay']).size().reset_index().rename(columns={0:'count'}).head(100)
# df_crsdep_dep.count()
# CRSDepTime does not seem to be set on some occations (set to 0). Means that when CRSDepTime is set to 0 and DepTime are equal the DepDelay might differ.
df_crsdep_dep[df_crsdep_dep['CRSDepTime'] == 0].count()

# Check to see if there might be a strong FD when CRSDepTime is actually set
df_crsdep_dep[df_crsdep_dep['CRSDepTime'] != 0].head(50)
# Sometimes DepTime - CRSDepTime != DepDelay even when CRSDepTime is set (example: row 2) -> Soft FD?

CRSDepTime  DepTime  DepDelay
0         1545   1549.0       4.0
1         1610   1610.0       0.0
2         1850   1905.0      15.0
3         1728   1727.0      -1.0
4         1030   1035.0       5.0
5         1048   1048.0       0.0
6         1440   1436.0      -4.0
7         1745   1745.0       0.0
8          640    640.0       0.0
9          605      NaN       NaN
10        1320   1321.0       1.0
11        1322   1320.0      -2.0
12         800    801.0       1.0
13        1045   1040.0      -5.0
14        1940   1940.0       0.0
15        1220   1225.0       5.0
16        1031   1119.0      48.0
17        1044   1116.0      32.0
18        1845   1843.0      -2.0
19        1612   1602.0     -10.0
20        1420   1439.0      19.0
21        2035   2308.0     153.0
22        1040   1040.0       0.0
23        1455   1508.0      13.0
24        1850   1850.0       0.0
25        1835   1832.0      -3.0
26         600    601.0       1.0
27        1410   1407.0      -3.0
28        1215   1222.0       7.0
29        1800   1802.0       2.0
30         930    928.0      -2.0
31        1245   1239.0      -6.0
32        1644   1647.0       3.0
33        1125   1131.0       6.0
34        2135   2143.0       8.0
35        1450   1448.0      -2.0
36        1005      NaN       NaN
37         600    556.0      -4.0
38         830    834.0       4.0
39         700    704.0       4.0
40         818    818.0       0.0
41        1125   1122.0      -3.0
42        2200   2159.0      -1.0
43        1900   1852.0      -8.0
44         745      NaN       NaN
45        1651   1651.0       0.0
46        1330   1540.0     130.0
47        1155   1152.0      -3.0
48        2040   2040.0       0.0
49         500    458.0      -2.0

# check Soft FD (ActualElapsedTime, TaxiIn, TaxiOut) -> (AirTime)

In [102]:
# ---------------- check Soft FD (ActualElapsedTime, TaxiIn, TaxiOut) -> (AirTime)---------------
df_air_time = df[['ActualElapsedTime', 'TaxiIn', 'TaxiOut', 'AirTime']]
df_air_time = df_air_time.sort_values(by=['ActualElapsedTime', 'TaxiIn', 'TaxiOut'])
df_air_time.head(50)

# Take a look at specific rows that have equal left hand side but AirTime is sometimes different.
df.iloc[[214815, 328863, 689521, 867367]]

# yet another few examples with same left hand side but sometimes different AirTime. Seems to be connection with CRSElapsedTime?
df.iloc[[119130, 615929, 673863]]


ActualElapsedTime  AirTime  ArrDelay  ArrTime  CRSArrTime  CRSDepTime  \
119130               16.0      9.0       7.0   1056.0        1049        1029   
615929               16.0     12.0      -7.0   1248.0        1255        1232   
673863               16.0      9.0      -2.0   1358.0        1400        1340   

        CRSElapsedTime CancellationCode  Cancelled  CarrierDelay  DayOfWeek  \
119130            20.0              NaN          0           NaN          1   
615929            23.0              NaN          0           0.0          5   
673863            20.0              NaN          0           NaN          5   

        DayofMonth  DepDelay  DepTime Dest  Distance  Diverted  FlightNum  \
119130          27      11.0   1040.0  PSG      31.0         0         65   
615929          14       0.0   1229.0  VIS      36.0         0       6055   
673863          15       2.0   1342.0  ITH      32.0         0       1175   

        LateAircraftDelay  Month  NASDelay Origin  SecurityDelay TailNum  \
119130                NaN      5       NaN    WRG            NaN  N747AS   
615929                0.0      5       0.0    FAT            0.0  N292UX   
673863                NaN     11       NaN    ELM            NaN  N886US   

        TaxiIn  TaxiOut UniqueCarrier  WeatherDelay  Year  
119130     3.0      4.0            AS           NaN  2002  
615929     3.0      4.0            OO           0.0  2004  
673863     3.0      4.0            US           NaN  1996

# check Soft FD (AirTime, TaxiIn, TaxiOut) -> (ActualElapsedTime)

In [134]:
df_air_time_delay = df[['AirTime', 'TaxiIn', 'TaxiOut', 'ActualElapsedTime']]
df_air_time_delay = df_air_time_delay.sort_values(by=['AirTime', 'TaxiIn', 'TaxiOut'])
df_air_time_delay[df_air_time_delay['AirTime'].notnull() & df_air_time_delay['TaxiIn'].notnull() & df_air_time_delay['TaxiOut'].notnull()].head(500)





AirTime  TaxiIn  TaxiOut  ActualElapsedTime
139136  -1419.0  1441.0     20.0               42.0
633330  -1417.0  1443.0     14.0               40.0
708954  -1414.0  1444.0      7.0               37.0
773258  -1413.0  1442.0     13.0               42.0
140584  -1413.0  1444.0     17.0               48.0
464072  -1410.0  1444.0     15.0               49.0
211788  -1410.0  1444.0     16.0               50.0
323928  -1410.0  1445.0     20.0               55.0
951785  -1409.0  1442.0      8.0               41.0
922405  -1409.0  1445.0      8.0               44.0
664103  -1409.0  1445.0      9.0               45.0
519509  -1408.0  1443.0     32.0               67.0
527779  -1408.0  1445.0      7.0               44.0
278399  -1407.0  1442.0     21.0               56.0
89217   -1407.0  1443.0      5.0               41.0
177070  -1407.0  1444.0     18.0               55.0
663768  -1407.0  1445.0     15.0               53.0
986930  -1406.0     4.0      7.0               45.0
199485  -1405.0  1446.0      9.0               50.0
214731  -1405.0  1448.0     13.0               56.0
145230  -1404.0  1443.0      8.0               47.0
393140  -1404.0  1447.0      9.0               52.0
383600  -1402.0  1443.0      6.0               47.0
34421   -1401.0  1442.0      6.0               47.0
141942  -1401.0  1443.0      5.0               47.0
321925  -1401.0  1444.0     16.0               59.0
587003  -1400.0  1442.0     23.0               65.0
355363  -1400.0  1443.0     10.0               53.0
902390  -1400.0  1443.0     40.0               83.0
741063  -1399.0  1444.0     12.0               57.0
755259  -1398.0  1447.0     17.0               66.0
778077  -1398.0  1448.0     29.0               79.0
739843  -1397.0  1444.0     13.0               60.0
10817   -1396.0  1442.0      5.0               51.0
794970  -1396.0  1442.0     16.0               62.0
886951  -1396.0  1445.0     21.0               70.0
693926  -1395.0  1442.0     10.0               57.0
447073  -1394.0  1460.0      7.0               73.0
39794   -1393.0  1442.0      5.0               54.0
907170  -1390.0  1442.0      5.0               57.0
103919  -1390.0  1443.0     18.0               71.0
507206  -1389.0  1446.0     12.0               69.0
533288  -1386.0  1444.0     19.0               77.0
890699  -1386.0  1445.0     11.0               70.0
206561  -1385.0  1439.0     13.0               67.0
363706  -1385.0  1445.0     20.0               80.0
686174  -1384.0  1444.0     13.0               73.0
385502  -1383.0  1445.0     13.0               75.0
599409  -1383.0  1447.0     26.0               90.0
980198  -1382.0  1442.0     18.0               78.0
629586  -1382.0  1443.0     13.0               74.0
175041  -1381.0  1442.0     44.0              105.0
245022  -1381.0  1444.0     19.0               82.0
741659  -1381.0  1444.0     23.0               86.0
717380  -1381.0  1444.0     33.0               96.0
171267  -1380.0  1442.0      6.0               68.0
902222  -1379.0  1443.0      6.0               70.0
760701  -1379.0  1445.0      9.0               75.0
908072  -1379.0  1445.0     14.0               80.0
290920  -1378.0  1439.0     14.0               75.0
968803  -1378.0  1442.0      6.0               70.0
892763  -1378.0  1447.0     16.0               85.0
470760  -1377.0  1439.0     23.0               85.0
573348  -1376.0  1442.0      6.0               72.0
912890  -1376.0  1443.0      9.0               76.0
242332  -1376.0  1446.0      4.0               74.0
991485  -1374.0  1445.0     13.0               84.0
507585  -1373.0  1446.0     37.0              110.0
743828  -1373.0  1448.0     45.0              120.0
780782  -1372.0  1442.0     10.0               80.0
861016  -1369.0  1443.0     13.0               87.0
491188  -1369.0  1444.0     19.0               94.0
682777  -1368.0     2.0   1417.0               51.0
600983  -1368.0  1443.0      7.0               82.0
76130   -1366.0  1444.0     12.0               90.0
882954  -1365.0  1444.0     13.0               92.0
4828

In [133]:
df_air_time_delay.groupby(['AirTime', 'TaxiIn', 'TaxiOut']).count().reset_index().rename(columns={0:'count'})


AirTime  TaxiIn  TaxiOut  ActualElapsedTime
0       -1419.0  1441.0     20.0                  1
1       -1417.0  1443.0     14.0                  1
2       -1414.0  1444.0      7.0                  1
3       -1413.0  1442.0     13.0                  1
4       -1413.0  1444.0     17.0                  1
...         ...     ...      ...                ...
130236    630.0     8.0     14.0                  1
130237    637.0     6.0     26.0                  1
130238    655.0     4.0     19.0                  1
130239   1562.0     4.0     66.0                  1
130240   1570.0    21.0      7.0                  1

[130241 rows x 4 columns]

In [132]:
df_air_time_delay.groupby(['AirTime', 'TaxiIn', 'TaxiOut', 'ActualElapsedTime']).count().reset_index().rename(columns={0:'count'})
# only +- 6.000 different groups between the two so soft dependency?

AirTime  TaxiIn  TaxiOut  ActualElapsedTime
0       -1419.0  1441.0     20.0               42.0
1       -1417.0  1443.0     14.0               40.0
2       -1414.0  1444.0      7.0               37.0
3       -1413.0  1442.0     13.0               42.0
4       -1413.0  1444.0     17.0               48.0
...         ...     ...      ...                ...
136064    630.0     8.0     14.0              652.0
136065    637.0     6.0     26.0              669.0
136066    655.0     4.0     19.0              678.0
136067   1562.0     4.0     66.0             1812.0
136068   1570.0    21.0      7.0             1538.0

[136069 rows x 4 columns]

In [135]:
# Take a look at specific rows that have equal left hand side but AirTime is sometimes different.
df.iloc[[494368, 508073, 645262, 968918]]

ActualElapsedTime  AirTime  ArrDelay  ArrTime  CRSArrTime  CRSDepTime  \
494368               17.0      9.0      64.0   1837.0        1733        1714   
508073               24.0      9.0       0.0   2143.0        2200        2136   
645262               17.0      9.0       6.0   1301.0        1255        1232   
968918               17.0      9.0      88.0   1200.0        1032        1010   

        CRSElapsedTime CancellationCode  Cancelled  CarrierDelay  DayOfWeek  \
494368            19.0              NaN          0           NaN          3   
508073            24.0              NaN          0           NaN          5   
645262            23.0              NaN          0           0.0          5   
968918            22.0              NaN          0           0.0          6   

        DayofMonth  DepDelay  DepTime Dest  Distance  Diverted  FlightNum  \
494368          17      66.0   1820.0  WRG      31.0         0         64   
508073          17       0.0   2126.0  SGU      50.0         0       3622   
645262           9      12.0   1244.0  VIS      36.0         0       6055   
968918           9      93.0   1143.0  PSG      31.0         0         65   

        LateAircraftDelay  Month  NASDelay Origin  SecurityDelay TailNum  \
494368                NaN      6       NaN    PSG            NaN  N744AS   
508073                NaN      1       NaN    CDC            NaN  N308SW   
645262                0.0      1       0.0    FAT            0.0  N292UX   
968918               88.0      7       0.0    WRG            0.0  N783AS   

        TaxiIn  TaxiOut UniqueCarrier  WeatherDelay  Year  
494368     3.0      5.0            AS           NaN  1998  
508073     3.0      5.0            OO           NaN  2003  
645262     3.0      5.0            OO           0.0  2004  
968918     3.0      5.0            AS           0.0  2005